arangod/Pregel/Algos/HITSKleinberg.cpp

```diff
@@ -0,0 +1,333 @@
+    double const hubMaxDiff = *getAggregatedValue<double>(maxDiffHubAggregator);
+    double const diff = std::max(authMaxDiff, hubMaxDiff);
+
+    if (diff < epsilon) {
+      if (state == State::updateAuthNormalizeHub) {
+        state = State::finallyNormalizeHubs;
+      } else if (state == State::updateHubNormalizeAuth) {
+        state = State::finallyNormalizeAuths;
+      }
+    }
+  }
+
+  void postGlobalSuperstep(uint64_t gss) override {
+    switch (state) {
+      case State::sendInitialHubs: {
+        state = State::updateAuth;
+      } break;
+      case State::updateAuth:
+      case State::updateAuthNormalizeHub: {
+        state = State::updateHubNormalizeAuth;
+      } break;
+      case State::updateHubNormalizeAuth: {
+        ++currentIteration;
+        if (currentIteration == numIterations) {
+          state = State::finallyNormalizeHubs;
+        } else {
+          state = State::updateAuthNormalizeHub;
+        }
+      } break;
+      case State::finallyNormalizeHubs:
+      case State::finallyNormalizeAuths: {
+        aggregate(isLastIterationAggregator, true);
+      } break;
+    }
+  }
+};
+
+struct HITSKleinbergComputation
+    : public VertexComputation<VertexType, int8_t, SenderMessage<double>> {
+  HITSKleinbergComputation() = default;
+
+  // sends auth to all in-neighbors. note that all vertices send messages
+  // to all out-neighbors in all iterations (there are no inactive vertices),
+  // so the set of in-neighbors can be determined by iterating over received
+  // messages
+  void sendAuthToInNeighbors(
+      MessageIterator<SenderMessage<double>> const& receivedMessages,
+      double auth) {
+    SenderMessage<double> authData(this->pregelId(), auth);
+    for (SenderMessage<double> const* message : receivedMessages) {
+      sendMessage(message->senderId, authData);
+    }
+  }
+
+  void sendHubToOutNeighbors(double hub) {
+    SenderMessage<double> hubData(this->pregelId(), hub);
+    sendMessageToAllNeighbours(hubData);
+  }
+
+  void compute(
+      MessageIterator<SenderMessage<double>> const& messages) override {
+    auto const* ctx =
+        dynamic_cast<HITSKleinbergWorkerContext const*>(context());
+    switch (ctx->state) {
+      case State::sendInitialHubs: {
+        // these are not normalized, but according to the description of the
+        // algorithm in the paper, the 1.0's are used in place of normalized
+        // values
+        mutableVertexData()->normalizedAuth = 1.0;
+        mutableVertexData()->normalizedHub = 1.0;
+        reportFakeDifference();
+
+        sendHubToOutNeighbors(1.0);
+      } break;
+
+      case State::updateAuth: {
+        // We enter this state when all authorities and all hubs are 1.0
+        updateStoreAndSendAuth(messages, 1.0);
+        reportFakeDifference();
+        // authorities are one iteration before hubs
+        // authorities are not normalized, hubs are 1.0
+
+        // Next state: updateHubNormalizeAuth
```

Comment:
just out of curiosity and because we discussed the state machine: Could you also switch the state here to have only one switch statement or is it that you want this separate function postGlobalSuperstep that only does the state changes?

Reply:
It would be possible. I didn't do this for two reasons. One is that I wanted to separate the logic of switching states from the logic of states implying certain behavior. (I am not sure that every new switch makes the code worse.) The more important one is that the state is global, i.e., (1) the same for all vertices and (2) stored only once in the context. So we change it only once per gss, in the Worker. (Storing the state in each vertex would require a lot of additional memory.)